### ISSUE TITLE
[MSHADE-223] Endless processing when promoteTransitiveDependencies=true

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-shade-plugin#584]
**[Fabrizio Cucci](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=fabrizio.cucci)** opened **[MSHADE-223](https://issues.apache.org/jira/browse/MSHADE-223?redirect=false)** and commented

With the following simple pom:

```xml
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>

	<groupId>com.coolworks</groupId>
	<artifactId>bug-test</artifactId>
	<version>1.6.1</version>

	<properties>
		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
		<maven.compiler.source>1.8</maven.compiler.source>
		<maven.compiler.target>1.8</maven.compiler.target>
	</properties>

	<dependencies>

		<dependency>
			<groupId>org.apache.spark</groupId>
			<artifactId>spark-core_2.10</artifactId>
			<version>1.6.1</version>
			<optional>true</optional>
		</dependency>

	</dependencies>

	<build>
		<plugins>
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-shade-plugin</artifactId>
				<version>2.4.3</version>
				<executions>
					<execution>
						<phase>package</phase>
						<goals>
							<goal>shade</goal>
						</goals>
						<configuration>
							<promoteTransitiveDependencies>true</promoteTransitiveDependencies>
							<artifactSet>
								<includes>
									<include>org.apache.spark:spark-core_2.10</include>
								</includes>
							</artifactSet>
						</configuration>
					</execution>
				</executions>
			</plugin>
		</plugins>
	</build>

</project>
```

running "mvn clean package" leads the plugin to endlessly print:

```
[INFO] Dependency-reduced POM written at: /Users/myuser/Documents/workspace/bug-test/dependency-reduced-pom.xml
```

---

**Affects:** 2.4.3

**Issue Links:**
- [MSHADE-225](https://issues.apache.org/jira/browse/MSHADE-225) Writing output only once
 (_**"is superceded by"**_)

**Remote Links:**
- [GitHub Pull Request #15
](https://github.com/apache/maven-shade-plugin/pull/15)

2 votes, 6 watchers

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Should be fixed as part of MSHADE-225

**Comment by jira-importer** [comment]
**[Fabrizio Cucci](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=fabrizio.cucci)** commented

Just to be clear: the issue was not about printing multiple times something but not terminating at all. I'll give it another try with the new snapshot version.

**Comment by jira-importer** [comment]
**[Henning Hoefer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hoeferh)** commented

While the output now (3.0.0) indeed appears only once, the build still isn't terminating.

Turning off the dependency-reduced off is my current workaround - but this issue **is not fixed**. Please reopen.

**Comment by jira-importer** [comment]
**[Bryan Harclerode](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=baharclerode)** commented

I ran into this when I pulled in two artifacts with the same GAV but different classifiers. Specifically, [this check](https://github.com/apache/maven-plugins/blob/trunk/maven-shade-plugin/src/main/java/org/apache/maven/plugins/shade/mojo/ShadeMojo.java#L1161) does not take into account classifiers, so it repeatedly adds the exclusion to the wrong dependency. If it's trying to exclude something from the second dependency, it will add it to the first dependency and iterate, detect that something still needs to be excluded from the second dependency (since it never got the exclusion) and repeat indefinitely.

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Fixed in [038d807b38115ff093bf16f3b0a2db439202bf67](https://gitbox.apache.org/repos/asf?p=maven-shade-plugin.git;a=commit;h=038d807b38115ff093bf16f3b0a2db439202bf67)
Thanks for the patch!
